flocker/node/agents/test/test_blockdevice.py

```diff
@@ -1198,16 +1205,146 @@ class DidNotConverge(Exception):
     """
 
 
-class BlockDeviceCalculatorTests(SynchronousTestCase):
+class _WriteVerifyingExternalClient(object):
     """
-    Tests for ``BlockDeviceCalculator``.
+    Representation of an external client of the Flocker API. This client can be
+    used to try to write to paths on the system and verify that the writes are
+    persisted to the underlying blockdevice.
+
+    :ivar _known_mountpoints: Known mountpoints of the dataset. Once flocker
+        has returned a path to an external agent, there is no promising how
+        long that agent might cache that information before it tries to write
+        to that location. This agent keeps track of all mountpoints it has ever
+        seen and tries to write to all of them.
+    :ivar _device_path: The device path of the blockdevice that backs the
```

Comment:
So I'll add a comment here, I think this actually has to be part of state. Otherwise after the dataset gets to ATTACHED I think the device path is removed from NodeState, so this test harness would have no hook to determine the write was persisted to a blockdevice or not without some knowledge of the fact that there is no way to get a write to a backing blockdevice if flocker isn't reporting a device in `NodeState`. This happens to be true right now, but is the sort of white-box assumption-based testing I try to avoid.